rfcs/0003-template-compiler.md

```diff
@@ -0,0 +1,127 @@
+- Provide a TypeScript transformer that can be used with a rollup plugin such as [@rollup/plugin-typescript](https://www.npmjs.com/package/@rollup/plugin-typescript).
+- Compiler can be run on JavaScript and TypeScript files.
+
+### Non-Goals
+
+- The compiled template does not have to be syntactically identical to the html prepared at runtime, only semantically identical.
+- Other template transformations. This is only for performance.
+
+## Motivation
+
+`lit-html` has [three internal rendering phases](https://github.com/lit/lit/blob/main/dev-docs/design/how-lit-html-works.md#summary-of-lit-html-rendering-phases) which are all optimized to cache and reuse previous work. However, the very first `render()` must still execute all three phases.
+
+The @lit-labs/compiler will provide a way to opt into precompiling the templates, moving the first internal rendering phase to build time.
+
+By precompiling `html` tagged template literals, the runtime only needs to execute two internal rendering phases reducing the first render performance. As a nice benefit, any subsequent updates that encounter a new compiled `html` tagged template literal will also benefit from the performance improvement and can skip the prepare phase.
+
+## Detailed Design
+
+At a high level, the `@lit-labs/compiler` package provides a transformer that converts:
+
+```js
+const hi = (name) => html`<h1>Hello ${name}!</h1>`;
+```
+
+into:
+
+```js
+const compiled_html = (s) => s;
+const lit_template_1 = { h: compiled_html`<h1>Hello <?></h1>`, parts: [{ type: 2, index: 1 }] };
+const hi = (name) => ({ _$litType$: lit_template_1, values: [name] });
+```
+
+This transform converts the `` html`<h1>Hello ${name}!</h1>` `` `TemplateResult` into a `CompiledTemplateResult` which semantically behaves the same but with greater initial render performance. The update performance is unchanged.
+
+### Transformer Design
+
+The transformer should be completely syntactic. Initially only compiling any `html` template expressions if `html` was imported from `lit` or `lit-html` directly.
+
+Each individual `TemplateResult` is split into two declarations, a `CompiledTemplate` and `CompiledTemplateResult`. The `CompiledTemplate` is hoisted to the module scope, and contains the prepared HTML and template parts for the original `TemplateResult`. The `CompiledTemplateResult` replaces the `TemplateResult` and binds the static `CompiledTemplate` with the dynamic `values` array.
+
+If a part is used that requires a constructor, the compiler
+will also insert an import such as:
+
+```js
+// ...
+import { _$LH as litHtmlPrivate_1 } from "lit-html/private-ssr-support.js";
+const { AttributePart: _$LH_AttributePart, PropertyPart: _$LH_PropertyPart, BooleanAttributePart: _$LH_BooleanAttributePart, EventPart: _$LH_EventPart } = litHtmlPrivate_1;
+```
+
+## Implementation Considerations
+
+### Implementation Plan
+
+Create a new `@lit-labs/compiler` package which exports a Typescript transformer.
```

Comment:
Is this compatible with the transformer from `@lit/localize-tools`?

There's the issue in `@lit/localize-tools` of the typescript version that localize-tools uses vs the version the user uses. Things may be broken if those do not match (because TS does breaking API changes in minor/patch versions....). Is the same issue applicable here?

The issue above stems from the fact that the *user* has a `rollup.config.js` where the transformers are included. Since that happens in "user space", it uses the typescript version of the user. It then passes the compiler objects to the Lit transformers. But if they don't use the same TS versions things may/can/will break. Not sure if this problem is solvable.

Reply:
I think you'll need to run `@lit-labs/compiler` after the localize transform. We should add a test for this.

TypeScript version compatibility will be an ongoing issue, as TypeScript makes breaking changes with each minor and major release. We'll want to document the TypeScript version(s) we support with each version of the compiler.

Reply:
Since identifying the tagged-template literals `html\`...\`` is done right now by string comparison of the name `html` which will need to be documented and/or configurable based on use of minifiers/other transformers.  (By not relying on the type system it speeds up the transform and helps decouple it from the TypeScript version, which tends to feature more updates to its type system in new versions than anything else.)

Reply:
As usergenic said above, the compiler has been designed so it works on standalone files, and doesn't require any TypeScript types. Thus the transformer can operate over JavaScript files.

An option we can explore that is maybe inelegant, but solves the TypeScript version issue, is to vend a separate rollup plugin that has all dependencies bundled. It would be large, but it would always work – and could easily be dropped into a rollup config to transform both TypeScript & JavaScript files.